File: binary_search/smaller_rights/smaller_rights.cpp
#include <vector>
using namespace std;

vector<int> countSmaller(vector<int> &nums)
{ // LeetCode Q.315.
    vector<int> smaller_rights, sorted_nums;
    for (int idx = nums.size() - 1; idx >= 0; idx--)
    {
        // Idx of the 1st right sorted num > current num - 1.
        int insertion_idx = upper_bound(sorted_nums.begin(), sorted_nums.end(), nums[idx] - 1) - sorted_nums.begin();

        sorted_nums.insert(sorted_nums.begin() + insertion_idx, nums[idx]);
        smaller_rights.push_back(insertion_idx);
    }
    reverse(smaller_rights.begin(), smaller_rights.end());
    return smaller_rights;
}

File: binary_search/smaller_rights/smaller_rights.py
from bisect import bisect_right


def count_smaller_rights(nums: list[int]) -> list[int]:  # LeetCode Q.315.
    smaller_rights, sorted_nums = [], []

    for num in nums[::-1]:  # Iteration: from rightmost to leftmost.
        # Idx of the 1st right side sorted num > current num - 1.
        idx = bisect_right(sorted_nums, num - 1)

        smaller_rights.append(idx)
        sorted_nums.insert(idx, num)

    return smaller_rights[::-1]  # Reverse to make order from leftmost to rightmost.
